### EDA - Combined Dateset

In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Reading the dataset
data = pd.read_csv('data/cleaned_data/combined_cleaned.csv')
data.head()b

domain  batch  \
0    Web Development Using Python Django  BCK01   
1  Web Development Using Node.js Express  BCK01   
2    Web Development Using Python Django  BCK01   
3                             MERN Stack  BCK01   
4    Web Development Using Python Django  BCK01   

                  placed_company         designation   age     degree cs_bg  \
0           VNC Digital Services   Software Engineer  22.0  No Degree    No   
1  Emstell Technology Consulting   Software Engineer  21.0     Degree    No   
2                 Creative Panda   Node JS Developer  22.0     Degree   Yes   
3              NORQ Technologies  Software Developer  23.0  No Degree    No   
4                       ActionFi   Software Engineer  22.0  No Degree    No   

  experience  avg_time holidays  week_back       city  \
0         No       9.0       No        0.0  Bangalore   
1         No      12.0       No        2.0        WFH   
2         No       9.0      Yes        4.0  Bangalore   
3         No      11.0      Yes        7.0    Chennai   
4         No      10.0       No        4.0  Bangalore   

                                              skills graduated  salary  
0  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.00  
1  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.16  
2  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40  
3  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.55  
4  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40

In [10]:
data.shape

(670, 15)

Dataset contains 670 rows and 15 columns

In [7]:
print(data.columns.values)

['domain' 'batch' 'placed_company' 'designation' 'age' 'degree' 'cs_bg'
 'experience' 'avg_time' 'holidays' 'week_back' 'city' 'skills'
 'graduated' 'salary']


In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 670 entries, 0 to 669
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   domain          670 non-null    object 
 1   batch           670 non-null    object 
 2   placed_company  670 non-null    object 
 3   designation     670 non-null    object 
 4   age             670 non-null    float64
 5   degree          670 non-null    object 
 6   cs_bg           670 non-null    object 
 7   experience      670 non-null    object 
 8   avg_time        670 non-null    float64
 9   holidays        670 non-null    object 
 10  week_back       670 non-null    float64
 11  city            670 non-null    object 
 12  skills          670 non-null    object 
 13  graduated       670 non-null    object 
 14  salary          670 non-null    float64
dtypes: float64(4), object(11)
memory usage: 78.6+ KB


In [18]:
# Dropping column batch
data = data.drop('batch', axis=1)

#### Descriptive Statistics

In [19]:
data.describe().T

count       mean       std   min    25%   50%   75%   max
age        670.0  22.323881  1.242447  19.0  22.00  22.0  23.0  31.0
avg_time   670.0  10.620896  1.985639   7.0   9.25  10.0  12.0  16.0
week_back  670.0   3.395522  2.413275   0.0   1.00   4.0   5.0   7.0
salary     670.0   4.534224  1.693921   1.8   3.60   4.3   5.0  30.0

* Average age of students in Brocamp is 22
* Min age is 19 --> Students joined staright after plus two
* Max age is 31 --> There are people from other professoins who is joining Brocamp
* Average time spent by students is 10.6 hours, which is higher than industrial average (8 hours)
* The least avg time is 7 --> Most possibly day scholars
* Max avg time recorde is 16 --> Students who are putting extra efforts - check corr with salary
* On an average 3 week backs are recieved to the sutdents
* There are students who got zero weekbacks
* Max salary students recieved is 30 LPA - (in UK)
* Min Salary is 1.8 LPA --> May be from early batches
* On an average students get placed with a salary of 4.5 LPA
* The age and salary seems to have right skewed distrubtion (outliers in rightside)

#### Missing values

In [20]:
# Checking for null values
(data.isnull().sum()/len(data))*100

domain            0.0
placed_company    0.0
designation       0.0
age               0.0
degree            0.0
cs_bg             0.0
experience        0.0
avg_time          0.0
holidays          0.0
week_back         0.0
city              0.0
skills            0.0
graduated         0.0
salary            0.0
dtype: float64

In [21]:
# Checking for duplicated values
data.duplicated().sum()

2

The dataset does not have any null values (which were treated in the previous stage) and neither does it have duplicated entries.

#### Data Quality Check

In [39]:
# Function for checking unique values in columsn
def get_unique(column):
    data[column] = data[column].str.lower()
    data[column] = data[column].str.strip()
    unique = data[column].value_counts()
    
    return unique

##### 1. Domain Column

In [27]:
get_unique('domain')

mern stack                                       341
mobile development using flutter                 132
web development using python django               76
web development using python django + react       37
mean stack                                        18
golang                                            14
mobile development using android kotlin            9
machine learning                                   9
web development using node.js express              9
devops                                             8
cybersecurity                                      7
mobile development using swift                     3
game development using unity                       2
data science                                       2
frontend development using react                   1
blockchain                                         1
web development using python django + angular      1
Name: domain, dtype: int64

##### 2. placed_company

In [24]:
data['placed_company'].nunique()

459

* Brocamp has placed students 459 companies
As this is not relevent to our problem statement we can drop this column

In [25]:
data = data.drop('placed_company', axis=1)
data.head()

domain         designation   age     degree  \
0    web development using python django   Software Engineer  22.0  No Degree   
1  web development using node.js express   Software Engineer  21.0     Degree   
2    web development using python django   Node JS Developer  22.0     Degree   
3                             mern stack  Software Developer  23.0  No Degree   
4    web development using python django   Software Engineer  22.0  No Degree   

  cs_bg experience  avg_time holidays  week_back       city  \
0    No         No       9.0       No        0.0  Bangalore   
1    No         No      12.0       No        2.0        WFH   
2   Yes         No       9.0      Yes        4.0  Bangalore   
3    No         No      11.0      Yes        7.0    Chennai   
4    No         No      10.0       No        4.0  Bangalore   

                                              skills graduated  salary  
0  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.00  
1  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.16  
2  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40  
3  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.55  
4  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40

##### 3. designation 

In [35]:
# Adjnustin pandas settings for printing full values
pd.set_option('display.max_rows', None)

print(get_unique('designation'))

# # Reset the display option to its default value
# pd.reset_option('display.max_rows')

full stack developer                          69
software engineer                             51
flutter developer                             44
software developer                            38
mern stack developer                          35
python developer                              14
flutter developer                             14
full stack developer                          12
associate software engineer                   11
software engineer                             10
junior software developer                     10
junior developer                               9
junior software engineer                       8
node js developer                              8
jr. software engineer                          7
trainee software engineer                      7
junior full stack developer                    7
front end developer                            6
junior flutter developer                       5
software engineer trainee                      5
react developer     

* We can see same entries in different formats, may be has extra whitespaces
* Some records are junior, while others has jr.

In [36]:
# Stripping values for removeing extra whitespaces
data['designation'] = data['designation'].str.strip()

print(get_unique('designation'))

full stack developer                         81
software engineer                            61
flutter developer                            58
software developer                           40
mern stack developer                         36
python developer                             16
associate software engineer                  15
junior software engineer                     13
junior software developer                    13
junior developer                             11
node js developer                            10
junior full stack developer                   9
associate product engineer                    8
junior flutter developer                      8
jr. software engineer                         8
trainee software engineer                     7
backend developer                             6
react developer                               6
front end developer                           6
software engineer trainee                     6
node.js developer                       

In [57]:
# Replaceing variations to correct format
replacements = {'jr. software engineer': 'junior software engineer', 'node.js developer ': 'nodejs developer', 'node js developer ': 'nodejs developer',
               'software engineer trainee': 'trainee software engineer', 'software development engineer-1': 'software development engineer',
               'mern developer': 'mern stack developer', 'full stack web developer': 'full stack developer', 'back end developer': 'backend developer',
               'full-stack developer': 'full stack developer', 'mern - stack developer': 'mern stack developer',
               'front-end developer': 'frontend developer', 'front end developer': 'frontend developer',
               'jr. software developer': 'junior software developer', 'sde-1': 'software developer',
               'fullstack developer': 'full stack developer', 'mobile application developer flutter': 'flutter developer',
               'flutter + firebase developer': 'flutter developer ', 'back-end developer': 'backend developer',
               'jr. react developer': 'junior react developer', 'trainee-software engineer': 'trainee software engineer',
               'software development engineer trainee': 'trainee software engineer', 'flutter developer (android/ios)': 'flutter developer',
               'trainee - software developer': 'trainee software developer', 'trainee-software developer': 'trainee software developer', 
               'software engineer(backend)': 'software engineer', 'software developer - trainee': 'trainee software developer',
               'jr.web developer': 'junior web developer', 'jr. react & python developer': 'junior fullstack python developer', 'trainee - software engineer': 'trainee software engineer',
               'oftware development engineer 1': 'software development engineer', 'software engineer (trainee)': 'trainee software engineer',
               'junior node js developer': 'junior nodejs developer', 'jr. app developer ': 'junior app developer',
               'software engineer 1': 'software engineer', 'jr. ml engineer': 'junior ml engineer', 'associate product engineer.': 'associated product engineer',
               'jr. developer': 'junior developer', 'flutter devoloper': 'flutter developer', 'jr. react js developer': 'junior react developer',
               'backend developer (django)': 'bakcend developer', 'fluttter developer': 'flutter developer', 'jr. full stack developer': 'junior full stack developer',
               'junior - full - stack developer': 'junior full stack developer', 'junior full-stack developer': 'junior full stack developer',
               'golang developer(member, technical staff)': 'golang devleoper', 'jr flutter developer': 'junior flutter developer', 'fullstack engineer': 'full stack engineer',
               'jr. flutter developer': 'junior flutter developer', 'junior software engineer(se 1)': 'junior software engineer',
               'golang devleoper': 'golang developer', 'fulll stack developer': 'full stack developer', 'jr node js developer': 'junior nodejs developer',
               'react js developer': 'react developer', 'flutter deevloper': 'flutter developer', 'node.js developer': 'node js developer',
               'jr software engineer': 'junior software engineer', 'bakcend developer': 'backend developer', 'jr. app developer': 'junior app developer',
               'reactjs developer': 'react developer', 'go developer': 'golang developer', 'junior software engineer (android)': 'junior software engineer',
               'jr. machine learning engineer': 'junior machine learning engineer', 'jr full stack python developer': 'junior full stack developer',
               'junior ml engineer': 'junior machine learning engineer', 'node js developer': 'nodejs developer', 'full stack developer': 'fullstack developer',
               'junior full stack developer': 'junior fullstack developer'}

data['designation'] = data['designation'].replace(replacements)
data['designation'].value_counts()

full stack developer                     90
flutter developer                        64
software engineer                        63
mern stack developer                     41
software developer                       41
junior software engineer                 26
nodejs developer                         18
trainee software engineer                18
junior software developer                16
python developer                         16
associate software engineer              15
junior fullstack developer               13
backend developer                        12
junior developer                         12
junior flutter developer                 10
frontend developer                        9
software development engineer             9
associate product engineer                8
react developer                           8
golang developer                          5
angular developer                         5
developer                                 5
web developer                   

##### 4. city

In [58]:
get_unique('city')

bangalore                    254
wfh                          235
manjeri malappuram            93
chennai                       20
calicut                       14
hyderabad                     11
on site                       11
nagercoil                      8
manjeri                        8
chennai works from office      5
ernakulam                      4
kochi                          2
office                         2
pune                           2
trivandrum                     1
Name: city, dtype: int64

* Bangalore is the city that has more placements, which has simlar number to work from home.
* Some entries as in different format like `manjeri`, and `manjeri malappuarm`
* Most of these data is imputed

In [60]:
replacements = {'manjeri malappuram': 'manjeri', 'on site': 'others', 'chennai works from office': 'chennai',
               'office': 'others'}

data['city'] = data['city'].replace(replacements)
data['city'].value_counts()

bangalore     254
wfh           235
manjeri       101
chennai        25
calicut        14
others         13
hyderabad      11
nagercoil       8
ernakulam       4
kochi           2
pune            2
trivandrum      1
Name: city, dtype: int64

* This is not a reliable columns, as `manjeri` shows high numbers, which is not true